File: examples/Csharp.ExampleApplication/Hooks/BatchMigrationCompleted/LogMigrationBatchesHook.cs
using System;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;
using Tableau.Migration;
using Tableau.Migration.Engine.Hooks;
using Tableau.Migration.Engine.Migrators.Batch;

namespace Csharp.ExampleApplication.Hooks.BatchMigrationCompleted
{
    #region class
    public class LogMigrationBatchesHook<T> : IContentBatchMigrationCompletedHook<T>
        where T : IContentReference
    {
        private readonly ILogger<LogMigrationBatchesHook<T>> _logger;

        public LogMigrationBatchesHook(ILogger<LogMigrationBatchesHook<T>> logger)
        {
            _logger = logger;
        }

        public Task<IContentBatchMigrationResult<T>?> ExecuteAsync(IContentBatchMigrationResult<T> ctx, CancellationToken cancel)
        {
            _logger.LogInformation(
                "{ContentType} batch of {Count} item(s) completed:{NewLine}{Statuses}",
                typeof(T).Name,
                ctx.ItemResults.Count,
                Environment.NewLine,
                String.Join(Environment.NewLine, ctx.ItemResults.Select(r => $"{r.ManifestEntry.Source.Location}: {r.ManifestEntry.Status}")));

            return Task.FromResult<IContentBatchMigrationResult<T>?>(ctx);
        }
    }
    #endregion
}

File: examples/Python.ExampleApplication/Hooks/batch_migration_completed/log_migration_batches_hook.py
import logging
from typing import TypeVar
from tableau_migration import(
    ContentBatchMigrationCompletedHookBase,
    IContentBatchMigrationResult,
    IUser
    )

T = TypeVar("T")

class LogMigrationBatchesHook(ContentBatchMigrationCompletedHookBase[T]):
    def __init__(self) -> None:
        super().__init__()
        self._logger = logging.getLogger(__name__)
        
    def execute(self, ctx: IContentBatchMigrationResult[T]) -> IContentBatchMigrationResult[T]:
        
        item_status = ""
        for item in ctx.item_results:
            item_status += "%s: %s".format(item.manifest_entry.source.location, item.manifest_entry.status)
                
        self._logger.info("%s batch of %d item(s) completed:\n%s", ctx._content_type, ctx.item_results.count, item_status)
        
        pass
    
class LogMigrationBatchesHookForUsers(ContentBatchMigrationCompletedHookBase[IUser]):
    def __init__(self) -> None:
        super().__init__()
        self._content_type = "User";

class LogMigrationBatchesHookForGoups(ContentBatchMigrationCompletedHookBase[IUser]):
    def __init__(self) -> None:
        super().__init__()
        self._content_type = "Group";